builtin/logical/mongodb/util.go

```diff
@@ -0,0 +1,134 @@
+package mongodb
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"net"
+	"net/url"
+	"strconv"
+	"strings"
+	"time"
+
+	"gopkg.in/mgo.v2"
+)
+
+// Unfortunately, mgo doesn't support the ssl parameter in its MongoDB URI parsing logic, so we have to handle that
+// ourselves. See https://github.com/go-mgo/mgo/issues/84
+func parseMongoURI(uri string) (*mgo.DialInfo, error) {
+	uinfo, err := extractMongoURL(uri)
+	if err != nil {
+		return nil, err
+	}
+	direct := false
```

Comment:
Rather than create all of these variables, please instantiate the `mgo.DialInfo` struct up top, which will automatically get default values, then simply fill in values as you range through the options.